src/WebJobs.Extensions.Http/HttpTriggerAttributeBindingProvider.cs

```diff
@@ -138,26 +138,43 @@ public async Task<ITriggerData> BindAsync(object value, ValueBindingContext cont
                 }
 
                 IValueProvider valueProvider = null;
-                object poco = null;
-                IReadOnlyDictionary<string, object> userTypeBindingData = null;
                 string invokeString = ToInvokeString(request);
                 if (_isUserTypeBinding)
                 {
                     valueProvider = await CreateUserTypeValueProvider(request, invokeString);
-                    if (_bindingDataProvider != null)
-                    {
-                        // some binding data is defined by the user type
-                        // the provider might be null if the Type is invalid, or if the Type
-                        // has no public properties to bind to
-                        poco = await valueProvider.GetValueAsync();
-                        userTypeBindingData = _bindingDataProvider.GetBindingData(poco);
-                    }
                 }
                 else
                 {
                     valueProvider = new HttpRequestValueBinder(_parameter, request, invokeString);
                 }
 
+                // collection of binding data is expensive, so we define a factory to allow this
+                // to happen lazily only when needed
+                Func<Task<IReadOnlyDictionary<string, object>>> bindingDataFactory = () =>
+                {
+                    return GetBindingDataAsync(request, valueProvider);
```

Comment:
Making binding data resolution lazy saves a bunch of work, since for http it involves attempted json deserialization of the requet body and a bunch of other operations, that usually aren't needed.